extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCacheImpl.java

```diff
@@ -89,6 +89,32 @@ public CompletionStage<V> get() {
                 });
     }
 
+    @Override
+    public <V> CompletableFuture<V> getIfPresent(Object key) {
+        Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED_MSG);
+        CompletableFuture<Object> existingCacheValue = cache.getIfPresent(key);
+
+        // record metrics, if not null apply casting
+        if (existingCacheValue == null) {
+            statsCounter.recordMisses(1);
+            return null;
+        } else {
+            LOGGER.tracef("Key [%s] found in cache [%s]", key, cacheInfo.name);
+            statsCounter.recordHits(1);
+
+            // cast, but still throw the CacheException in case it fails
+            return unwrapCacheValueOrThrowable(existingCacheValue)
+                    .thenApply(value -> {
+                        try {
+                            return (V) value;
+                        } catch (ClassCastException e) {
+                            throw new CacheException("An existing cached value type does not match the requested type", e);
+                        }
+                    });
```

Comment:
Sorry, I suggested something using a lambda in a previous comment but lambdas are actually forbidden in Quarkus runtime code.
```suggestion
                    .thenApply(new Function<>() {
                        @Override
                        public Object apply(Object value) {
                            try {
                                return (V) value;
                            } catch (ClassCastException e) {
                                throw new CacheException("An existing cached value type does not match the requested type", e);
                            }
                        }
                    });
```